### ISSUE TITLE
Using keras GRUCell layer in TFLM

### ISSUE THREAD

**Issue by yair-ehrenwald** [tensorflow/tensorflow#43380]
@tensorflow/micro

**System information**
- Host OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu
- TensorFlow installed from (source or binary): TFLM source, tf-nightly from installer
- Tensorflow version (commit SHA if source):  tf-nightly 2.4.0.dev20200918
- Target platform (e.g. Arm Mbed OS, Arduino Nano 33 etc.):

**Describe the problem**
My goal is to use a network using a keras.GRUCell layer in TFLM.

This issue is somwhat connected to:
https://github.com/tensorflow/tensorflow/issues/42582
and
https://github.com/tensorflow/tensorflow/issues/41690
As they both arose from the goal of using a GRU layer. 

This is a sample model:
state0 = tf.keras.Input(shape=( 1,128), dtype='float32', name='state0')
main_input = tf.keras.Input(shape=(1, 200), dtype='float32', name='main_input')
gru0, state0_out = tf.keras.layers.GRUCell(128,reset_after=True,recurrent_activation='sigmoid',use_bias = True,name="grucell0")(main_input,state0)

model3 = tf.keras.models.Model(inputs=[main_input,state0],outputs = [gru0,state0_out])

after adding in SPLIT_V, I ran into two problems:
A. Missing SHAPE operator - I will open a PR shortly.

B. A problem with tensor allocation which I will describe and suggest a workaround for TFLM and open a PR if you think the solution is acceptable.

**Please provide the exact sequence of commands/steps when you ran into the problem**

I converted the above model to TFLite and then imported into TFLM.
The error I got was from AllocationInfoBuilder::AddTensors() in micro_allocator.cc

"Logic error in memory planner, tensor 14 has an invalid lifetime: "
          "first_created: 1, last_used: -1",

I examined the ops the converter generated:
{'index': 0, 'op_name': 'SHAPE', 'inputs': array([0], dtype=int32), 'outputs': array([12], dtype=int32)
 **{'index': 1, 'op_name': 'UNPACK', 'inputs': array([12], dtype=int32), 'outputs': array([13, 14, 15], dtype=int32)**
 {'index': 2, 'op_name': 'RESHAPE', 'inputs': array([0, 3], dtype=int32), 'outputs': array([16], dtype=int32)
 {'index': 3, 'op_name': 'PACK', 'inputs': array([13,  5,  6], dtype=int32), 'outputs': array([17], dtype=int32)
 {'index': 4, 'op_name': 'FULLY_CONNECTED', 'inputs': array([16, 10, -1], dtype=int32), 'outputs': array([18], dtype=int32)
 {'index': 5, 'op_name': 'RESHAPE', 'inputs': array([18, 17], dtype=int32), 'outputs': array([19], dtype=int32)
 {'index': 6, 'op_name': 'ADD', 'inputs': array([19,  9], dtype=int32), 'outputs': array([20], dtype=int32)
 {'index': 7, 'op_name': 'SPLIT', 'inputs': array([ 7, 20], dtype=int32), 'outputs': array([21, 22, 23], dtype=int32)
 {'index': 8, 'op_name': 'RESHAPE', 'inputs': array([1, 4], dtype=int32), 'outputs': array([24], dtype=int32)
 {'index': 9, 'op_name': 'FULLY_CONNECTED', 'inputs': array([24, 11, -1], dtype=int32), 'outputs': array([25], dtype=int32)
 {'index': 10, 'op_name': 'SHAPE', 'inputs': array([1], dtype=int32), 'outputs': array([26], dtype=int32)
 {'index': 11, 'op_name': 'UNPACK', 'inputs': array([26], dtype=int32), 'outputs': array([27, 28, 29], dtype=int32)
 {'index': 12, 'op_name': 'PACK', 'inputs': array([27,  5,  6], dtype=int32), 'outputs': array([30], dtype=int32)
 {'index': 13, 'op_name': 'RESHAPE', 'inputs': array([25, 30], dtype=int32), 'outputs': array([31], dtype=int32)
 {'index': 14, 'op_name': 'ADD', 'inputs': array([31,  9], dtype=int32), 'outputs': array([32], dtype=int32)
 {'index': 15, 'op_name': 'SPLIT_V', 'inputs': array([32,  2,  7], dtype=int32), 'outputs': array([33, 34, 35], dtype=int32)
 {'index': 16, 'op_name': 'ADD', 'inputs': array([21, 33], dtype=int32), 'outputs': array([36], dtype=int32)
 {'index': 17, 'op_name': 'LOGISTIC', 'inputs': array([36], dtype=int32), 'outputs': array([37], dtype=int32)
 {'index': 18, 'op_name': 'SUB', 'inputs': array([ 8, 37], dtype=int32), 'outputs': array([38], dtype=int32)
 {'index': 19, 'op_name': 'ADD', 'inputs': array([22, 34], dtype=int32), 'outputs': array([39], dtype=int32)
 {'index': 20, 'op_name': 'LOGISTIC', 'inputs': array([39], dtype=int32), 'outputs': array([40], dtype=int32)
 {'index': 21, 'op_name': 'MUL', 'inputs': array([40, 35], dtype=int32), 'outputs': array([41], dtype=int32)
 {'index': 22, 'op_name': 'ADD', 'inputs': array([23, 41], dtype=int32), 'outputs': array([42], dtype=int32)
 {'index': 23, 'op_name': 'TANH', 'inputs': array([42], dtype=int32), 'outputs': array([43], dtype=int32)
 {'index': 24, 'op_name': 'MUL', 'inputs': array([38, 43], dtype=int32), 'outputs': array([44], dtype=int32)
 {'index': 25, 'op_name': 'MUL', 'inputs': array([37,  1], dtype=int32), 'outputs': array([45], dtype=int32)
 {'index': 26, 'op_name': 'ADD', 'inputs': array([45, 44], dtype=int32), 'outputs': array([46], dtype=int32)}

Notice that UNPACK uses tensors 14 and 15 as outputs but they aren't used later on in the model. This is also true for the next SHAPE/UNPACK pattern and tensors 28 and 29.

I assume this is caused for reasons unknown by the TFLite conveter? The model works fine in python so I assume the tensor allocator in TFLite doesn't check for this.

What happens in TFLM is that those tensors aren't "caught" in the flow marking last usage.

The solution/workaround I implemented and works for me was to initialize all tensors as persistent  (in the same way the current code does for the graph's output tensors):

for (int i = (subgraph->operators()->size() - 1); i >= 0; --i) {
    const auto* op = subgraph->operators()->Get(i);
    for (size_t n = 0; n < op->outputs()->size(); ++n) {
      const int tensor_index = op->outputs()->Get(n);
      AllocationInfo* current = &info_[tensor_index];
      current->last_used = subgraph->operators()->size() - 1;
    }
  }

Afterwards, the current implementation will handle all the tensors that serve as input to ops correctly and leave these problematic ones as persistent to the end of the invocation.

Another option would be to search specifically for this case and handle it separately and
YET another option would be to see what in the TFLite converter is doing this...

@advaitjain Please let me know what you think, I'd really like to get this sorted out :)

**Comment by yair-ehrenwald** [comment]
@advaitjain 

Opened a PR with what I described above as a possible fix:
https://github.com/tensorflow/tensorflow/pull/43382

Let me know what you think. I'd be glad to implement changes if you think this might break something or if there's a method you'd prefer.

**Comment by advaitjain** [comment]
Thanks, will take a proper look on Monday.

For the SHAPE pull request, can you open a separate one for any changes to
flatbuffer_conversions? That can be approved right away but sometimes take
longer to merge.

**Comment by yair-ehrenwald** [comment]
Thanks, will do.

Regarding the memory_allocator, I think from a memory footprint POV it would be better to search explicitly for this case and set the life time to the minimum required (creation and end of life at the same index) but thought such a test might "legalize" some cases that should genuinely raise a flag..

**Comment by advaitjain** [comment]
> Thanks, will do.
> 
> Regarding the memory_allocator, I think from a memory footprint POV it would be better to search explicitly for this case and set the life time to the minimum required (creation and end of life at the same index) but thought such a test might "legalize" some cases that should genuinely raise a flag..

Our suggestion would be to do some flatbuffer manipulations from python and leave the TFLM code unchanged. I'll send you some pointers later today.

**Comment by yair-ehrenwald** [comment]
> > Thanks, will do.
> > Regarding the memory_allocator, I think from a memory footprint POV it would be better to search explicitly for this case and set the life time to the minimum required (creation and end of life at the same index) but thought such a test might "legalize" some cases that should genuinely raise a flag..
> 
> Our suggestion would be to do some flatbuffer manipulations from python and leave the TFLM code unchanged. I'll send you some pointers later today.

Sounds interesting, thanks!

**Comment by advaitjain** [comment]
Update here is that our suggestion would involve using the python flatbuffer API to make some changes.

I'm in the process of writing some documentation that would help but it appears that we are going to hit the same underlying issue as #41846 -- i.e. if schema_py_generated.py is empty with the bazel builds, you will not be able to directly do what I have in mind.

I'm going to do some more investigation to see if there is some workaround here.

**Comment by pkgoogle** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The TFLite team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the TFLite space.

Thanks.
